### PR TITLE
My methods

### PR DESCRIPTION
**Pull request by alexandermorales-dev**
Ruby built-in methods were replicated using custom methods created from scratch

### CODE REVIEW

File: script.rb [hunk 1]
```diff
@@ -0,0 +1,102 @@
+module Enumerable
+  # my_each
+  def my_each
+    i = 0
+    while i < to_a.length
+      yield to_a[i]
+      i += 1
+    end
+    self
+  end
```
Comment (MkrtichSargsyan) [inline script.rb]: - [x] enumerables #my_each returns an Enumerator if no block is given (return to_enum(:my_each) unless block_given?)

File: script.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+  # my_each_with_index
+  def my_each_with_index
+    i = 0
+    while i < to_a.length
+      yield to_a[i], index(to_a[i])
+      i += 1
+    end
+    self
+  end
```
Comment (MkrtichSargsyan) [inline script.rb]: - [x] should have working ```my_each_with_index```  identical and returns the same thing as ruby's each_with_index. ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))

-  enumerable ```my_each_with_index``` returns an enumerator if no block is given
- It should work with ranges.
- It should work with hashes.

File: script.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+  # my_select
+  def my_select
+    my_arr = []
+    to_a.my_each do |num|
+      my_arr.push(num) if yield num
+    end
+    my_arr
+  end
```
Comment (MkrtichSargsyan) [inline script.rb]: - enumerables ```my_select``` returns an enumerator if no block is given

File: script.rb [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+  # my_all
+  def my_all
+    new_arr = []
+    to_a.my_each do |num|
+      new_arr.push(num) if yield num
+    end
+    new_arr.length == to_a.length
+  end
```
Comment (MkrtichSargsyan) [inline script.rb]: - [x] should have working ```my_all?```  identical and returns the same thing as ruby's all?. ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))

-  when no block or argument is given returns true when none of the collection members are false or nil
-  when no block or argument is given returns false when one of the collection members is false or nil
- when a class is passed as an argument returns true if all of the collection is a member of such ```class::Integer```
- when a class is passed as an argument returns false if any of the collection is not a member of such class
- when a Regex is passed as an argument returns true if all of the collection matches the Regex

File: script.rb [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+  # my_any
+  def my_any
+    new_arr = []
+    to_a.my_each do |num|
+      new_arr.push(num) if yield num
+    end
+    new_arr.length >= 1
+  end
```
Comment (MkrtichSargsyan) [inline script.rb]: - [x] should have working ```my_any?```  identical and returns the same thing as ruby's any?. ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))


-  when no block or argument is given returns true if at least one of the collection is not false or nil
- Your method doesn't consider the possibility when a class (String, Numeric, etc) is given as an argument ```some_enumerable.my_any?(String)```.

File: script.rb [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+  # my_none
+  def my_none
+    new_arr = []
+    to_a.my_each do |num|
+      new_arr.push(num) if yield num
+    end
+    new_arr.length.zero?
+  end
```
Comment (MkrtichSargsyan) [inline script.rb]: - [x] should have working ```my_none?```  identical and returns the same thing as ruby's none? ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))

- when no block or argument is given returns true only if none of the collection members is true
- when a class is passed as an argument returns true if none of the collection is a member of such class
- It should work with ranges.
- Your method doesn't consider the possibility when a class (String, Numeric, etc) is given as an argument ```some_enumerable.my_none?(String)```.
- Your method doesn't consider the possibility when a Regex is given as an argument (some_enumerable.my_none?(/w/)).

File: script.rb [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  # my_map
+  def my_map_proc(proc = nil)
+    new_arr = []
+    if proc.nil?
+      to_a.my_each { |num| new_arr.push(yield(num)) }
+    else
+      to_a.my_each { |num| new_arr.push(proc.call(num)) }
+    end
+    new_arr
+  end
```
Comment (MkrtichSargsyan) [inline script.rb]: - [x] should have working ```my_map_proc```  identical and returns the same thing as ruby's ```map```. ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))

- Your method should return an enumerator if no block or argument is given.

File: script.rb [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+      accum = init_value
+      i = 0
+    end
+    while i < to_a.length
+      accum = yield accum, self[i]
+      i += 1
+    end
+    accum
+  end
+end
```
Comment (MkrtichSargsyan) [inline script.rb]: - [x] should have working ```my_inject```  identical and returns the same thing as ruby's ```inject```. ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))


- Your method should work with symbols as arguments (:+) and perform the right operation.
- It should take a block as an argument and apply it to each element of the object

**Review by MkrtichSargsyan — CHANGES_REQUESTED** [review]
## _Status:_ Changes Requested :recycle:

## Hello Alexander Morales

Good job on this project! :+1:

1. Your README file looks professional
2. You have descriptive git commit messages.


However, there are some changes that you need to make in order to get your project approved.
Please take a look at the comments I left in your code.

- [x] Kindly add the question mark after ```my_any?, my_none?, my_all?``` functions


### Please, make the REQUIRED changes and submit your request again! :grinning:
### Good luck. 

> ps: . If you have any questions about my review or need help, you can message me on slack ( @mkrtichsargsyan )

File: script.rb [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+    x = to_a if to_a.is_a? Range
+    x = self if to_a.is_a? Array
+    i = 0
+    while i < x.length
+      yield x[i], i
+      i += 1
+    end
+    self
+  end
```
Comment (Berabjesus) [inline script.rb]: - [x] it should returns the range itself after calling the given block once for each element in self. Your my_each_with_index currently does not take a range and a hash.
```
  range = Range.new(5, 50) 
  my_each_with_index_output = ''
  block = proc { |num, idx| my_each_with_index_output + "Num: #{num}, idx: #{idx}\n" }

  p range.each_with_index(&block)
  p range.my_each_with_index(&block)

# The above result should be the same.
```

File: script.rb [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+      to_a.my_each { |num| return false if num == false || num.nil? }
+    elsif !block_given? && (param.is_a? Class)
+      to_a.my_each { |item| return false if item.class != param }
+    elsif !block_given? && (param.is_a? Regexp)
+      to_a.my_each { |items| return false unless items.match(param) }
+    elsif block_given?
+      to_a.my_each { |num| return false unless yield num }
+    end
+    true
+  end
```
Comment (Berabjesus) [inline script.rb]: - [x] returns true if all of the collection is a member of such class::Numeric
```
  p [1,2,3,4].all?(Numeric)
  p [1,2,3,4].my_all?(Numeric) 

# This should return true
```

- [x]  when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

```
  p [3,3,3,3, 2].all?(3) // false
  p [3,3,3,3,2].my_all?(3) // true

# This should be false
```

File: script.rb [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+      to_a.my_each { |num| return true if num }
+      return false
+    elsif param.is_a?(Class)
+      to_a.my_each { |num| return true if num.instance_of?(param) }
+      return false
+    else
+      to_a.my_each { |num| return true if yield num }
+    end
+    false
+  end
```
Comment (Berabjesus) [inline script.rb]: - [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
```
  p [1,2,3,4].any?(Numeric)
  p [1,2,3,4].my_any?(Numeric) 

 # The result should be the same.
```

- [x]  when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern. 
- [x] Check for some regex values too. Its currently throwing a Localjumperror 
```
  p ['cat', 'dog'].any?('cat')
  p ['cat', 'dog'].my_any?('cat')

# this should be true
```

File: script.rb [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+      to_a.my_each { |num| return false if yield num }
+      return true
+    elsif !block_given? && param.instance_of?(Regexp)
+      to_a.my_each { |item| return false if item.match(param) }
+      return true
+    else
+      to_a.my_each { |num| return false if yield num }
+    end
+    true
+  end
```
Comment (Berabjesus) [inline script.rb]: - [x]  when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern.  Its currently throwing a Localjumperror.
```
 p ['cat', 'dog'].none?('bird')
 p ['cat', 'dog'].my_none?('bird')

# This should be true
```

**Review by Berabjesus — CHANGES_REQUESTED** [review]
## Status: CHANGES REQUIRED :warning: :construction: 
Hi @alexandermorales-dev  great job on working on the comments given by the earlier reviewer. 
#### However you still do have a few changes to carry out before we proceed so please take a look at the things you have to work on and their respective review comments below.
- [x] should have working `my_each_with_index` identical and returns the same thing as ruby's each_with_index. ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- [x] should have working `my_all?` identical and returns the same thing as ruby's `all?`. ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- [x] should have working `my_any?`  identical and returns the same thing as ruby's `any?`. ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- [x] should have working `my_none?` identical and returns the same thing as ruby's `none?`. ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))

#### Check the comments below.

#### **If you have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help** 
**HAPPY CODING :computer:[Bereket Beshane](https://bereketababu.me)** | [LinkedIn](https://linkedin.com/in/bereketbeshane) | [GitHub](https://github.com/Berabjesus)

**Review by SajjadAhmad14 — CHANGES_REQUESTED** [review]
# Changes Required :recycle:
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hello, you have been doing a tremendous job so far :ok_hand:
## You have worked on most of the changes suggested previously.
## But there are some more changes to make in order to get your project approved.

### #my_any
Please see [docs](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F) for reference and make it work exactly the same as #none? does.

- [x] It should return true when called on an empty array, but it's returning false.
- [x]  It should return false when at least one element is true. It's returning true at the moment. Kindly fix that!

### Happy Coding! 

## You can always connect with me on slack ( @Sajjad Ahmad ).

**Review by mricanho — APPROVED** [review]
# Status: ✔️ Project Approved ✔️ 

Hello @alexandermorales-dev, good job on yours enumerables! 🔥 

You worked hard in all the instructions that the previous Code Reviewers give you👍🏽 

ok ok


Now you can merge this branch 🚀 

You can contact me on slack if you have any questions, just search for me as @ Miguel Ricaño.

### 💻 👩🏽‍💻  Happy Coding!  👨🏽‍💻 🖥️

**Pull request merged**